### PR TITLE
Allowing schedule(pause/resume/unlayout) to go deep into the subtree.

### LINKED ISSUE #4503: Videos In Amp-Carousel In Amp Lightbox Don't Stop On Close
On pages in which there is a video inside a carousel inside a lightbox, the videos are not automatically paused when the lightbox is closed. Conversely, If the videos are not inside a carousel, they are paused, so it seems to be an issue with the nesting of carousels and lightboxes that breaks the pause-on-close behavior.

You can reproduce this issue at this [JSBin page](http://jsbin.com/sewanarura/14/edit?html,output). 
1. Click the "open lightbox" button
2. Play the youtube video
3. Click the "close lightbox" button
4. Observe that the audio is still playing from the youtube video

Affects AMP version 1471018286527 running in Chrome.

### PR DESCRIPTION
**Pull request by aghassemi**
Fixes #4503 

Currently when an element is scheduled for pause/resume/unlayout, only its direct amp children are called. The issue is that if an amp-element lives inside another amp-element, it will not get the scheduled events ( root cause of #4503 where there is an `amp-youtube` inside an `amp-carousel` inside an `amp-lightbox` ).

With this PR, we allow pause/resume/unlayout to go to _all_ descendants and not just the shallow subtree.

### CODE REVIEW

**Comment by erwinmombay** [comment]
image

File: src/service/resources-impl.js [hunk 1]
```diff
@@ -506,7 +508,8 @@ export class Resources {
     const parentResource = Resource.forElement(parentElement);
     subElements = elements_(subElements);
 
-    this.discoverResourcesForArray_(parentResource, subElements, resource => {
+    this.discoverResourcesForArray_(parentResource, subElements,
+    /* deepSubtree */ true, resource => {
```
Comment (jridgewell) [inline src/service/resources-impl.js]: Shouldn't the parent be in charge of unlayout-ing it's children? We have it being responsible for the layout.

File: src/service/resources-impl.js [hunk 2]
```diff
@@ -1417,36 +1421,51 @@ export class Resources {
...
       resource.setInViewport(inViewport);
     });
   }
 
   /**
-   * Finds resources within the parent resource's shallow subtree.
+   * Finds resources within the parent resource's shallow or deep subtree.
    * @param {!Resource} parentResource
    * @param {!Array<!Element>} elements
+   * @param {!boolean} deepSubtree Whether the parent resource's shallow or deep
```
Comment (jridgewell) [inline src/service/resources-impl.js]: `{boolean}`

File: src/service/resources-impl.js (hunk 1 above)
Comment (aghassemi) [inline src/service/resources-impl.js]: To be symmetric with `layout`, yes you are correct but I think the difference is that laying out children requires custom component logic where as unlayout doesn't. I am open to not doing this for `unlayout` but I think doing it is the safer approach to eliminate bugs like #4503.

Comment (dvoytenko) [inline src/service/resources-impl.js]: So, I'd say, `deepSubtree` should not go into `owned` elements.

That being said, I'd like to discuss this offline - I think we just need to go ahead and start working on layers. Possibly we can squeeze this PR in for a temporary fix, but I'd like to make sure we are not digging a deeper hole.

**Comment by mrjoro** [comment]
PR #4647 mentioned that it supersedes this one.

**Pull request closed without merging**